src/main/scala/mesosphere/marathon/core/storage/store/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -159,3 +161,224 @@ class LazyCachingPersistenceStore[K, Category, Serialized](
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings(Array("all")) // async/await
+  override def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized]): Future[Done] = {
+
+    if (!ir.hasVersions)
+      store.store(id, v)
+    else {
+      withVersionCache(id) { (category, storageId) =>
+        withVersionedValueCache {
+          async { // linter:ignore UnnecessaryElseBranch
+            await(store.store(id, v))
+            val version = ir.version(v)
+            updateCachedVersions(storageId, version, category, v)
+            Done
+          }
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings(Array("all")) // async/await
+  override def store[Id, V](id: Id, v: V, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized]): Future[Done] = {
+
+    withVersionCache(id) { (category, storageId) =>
+      withVersionedValueCache {
+        async { // linter:ignore UnnecessaryElseBranch
+          await(store.store(id, v, version))
+          updateCachedVersions(storageId, version, category, v)
+          Done
+        }
+      }
+    }
+  }
+
+  override def deleteCurrent[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] =
+    deleteCurrentOrAll(k, () => store.deleteCurrent(k))
+
+  override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] =
+    deleteCurrentOrAll(k, () => store.deleteAll(k))
+
+  override def deleteVersion[Id, V](
+    k: Id,
+    version: OffsetDateTime)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] =
+    deleteCurrentOrAll(k, () => store.deleteVersion(k, version))
+
+  override def versions[Id, V](id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] = {
+    val versionsFuture = withVersionCache(id) { (category, storageId) =>
+      if (versionCache.contains((category, storageId))) {
+        Future.successful(versionCache((category, storageId)))
+      } else {
+        async { // linter:ignore UnnecessaryElseBranch
+          val children = await(store.versions(id).toMat(Sink.seq)(Keep.right).run())
+          versionCache((category, storageId)) = children
+          children
+        }
+      }
+    }
+    Source.fromFuture(versionsFuture).mapConcat(identity)
+  }
+
+  override def ids[Id, V]()(implicit ir: IdResolver[Id, V, Category, K]): Source[Id, NotUsed] = store.ids()
+
+  override def storageVersion(): Future[Option[StorageVersion]] = store.storageVersion()
+
+  override def setStorageVersion(storageVersion: StorageVersion): Future[Done] =
+    store.setStorageVersion(storageVersion)
+
+  protected[cache] def maybePurgeCachedVersions(
+    maxEntries: Int = config.maxEntries,
+    purgeCount: Int = config.purgeCount,
+    pRemove: Double = config.pRemove): Unit =
+
+    while (versionedValueCache.size > maxEntries) {
+      // randomly GC the versions
+      val counter = new AtomicInteger
```

Comment:
This doesn't need to be atomic, its totally local to the function and completely unshared.